[CU-f92zhu] Create tournaments list view

eastemployeer opened the pull request.

Buglik committed:
init comit

Buglik committed:
init comit

Buglik committed:
TournamentsList starting point

Buglik committed:
TournamentsList starting point

Buglik committed:
mobile version (work in progress)

Buglik committed:
mobile version (work in progress)

Buglik committed:
improved mobile version
deleted pagination,
nested list items in btn,
added green/orange status color,
added sticky addBtn,
commented Participants section,

Buglik committed:
improved mobile version
deleted pagination,
nested list items in btn,
added green/orange status color,
added sticky addBtn,
commented Participants section,

eastemployeer committed:
Added desktop list tournament view

eastemployeer committed:
Added desktop list tournament view

eastemployeer committed:
Fixed tournament structure

eastemployeer committed:
Fixed tournament structure

Zummek commented:
Task linked: [CU-f92zhu Create tournaments list view](https://app.clickup.com/t/f92zhu)

Zummek commented on web-app/src/router/routes.ts:
```diff
@@ -6,10 +6,15 @@ const routes: RouteConfig[] = [
     children: [
       {
-        path: '',
+        path: '/index',
```
```suggestion
        path: '',
```
Under `/app` or `/app/tournament(s)` path we should have a tournament list view
`/app/tournament/:id` - details
`/app/tournament/creator` - creator

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+        grid
+        :data="data"
+        :columns="columns"
+        row-key="name"
```
```suggestion
        row-key="id"
```

Zummek replied:
UP!

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+      >
+        <template v-slot:top-left>
+          <div class="row">
+            <span class="title">Tournaments</span>
```
center vertically this txt

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+            debounce="300"
+            v-model="filter"
+            placeholder="Search"
+            style="width:150px"
```
`style="width:150px"` it's not bad but why did You set this width. In original width should be about a 190px and in my opinion this looks a bit better

eastemployeer replied:
Ask @Buglik

Buglik replied:
i did so 'cause it looked better at mobile view
probably @eastemployeer just copied it when preparing desktop but i agree with @Zummek , default looks better on desktop

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+        <template v-slot:item="props">
+          <div class="q-pa-sm col-sm-12 col-md-6 col-lg-4">
+            <q-card class="flex row">
+              <q-card-section class="col-4 overflow-hidden card__logo">
+                <q-icon size="5em" name="emoji_events" />
+              </q-card-section>
+
+              <q-card-section class="col-8 overflow-hidden">
```
`overflow-hidden` is used for?

eastemployeer replied:
Ask @Buglik

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+                </q-chip>
+                <q-avatar size="2em" v-for="n in 10" :key="n" class="q-ml-xs">
+                  <img src="https://cdn.quasar.dev/img/boy-avatar.png" />
+                </q-avatar>
```
UUU i have same idea. Maybe show/hide full names on hover.  `Organizer` will be name of user e.g. dawid kwiatkowski :P? But this is a detail

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+  justify-content: center;
+  align-items: center;
+}
+.card__header {
+  font-size: 30px;
+  font-weight: 400;
+  padding: 0;
+  margin: 0;
+  text-align: center;
+}
```
Dont use doubled floor in custom names. Names like this are in quasar, when you want to overwrite this class maybe you should create your own class or use seconds to overwrite the first

eastemployeer replied:
Google -> BEM metodology :))

Zummek commented on web-app/src/components/tournament/GridListDesktop.vue:
```diff
@@ -0,0 +1,110 @@
+}
+</script>
+<style lang="scss" scoped>
+.card__logo {
+  display: flex;
+  justify-content: center;
+  align-items: center;
+}
```
Same as below

eastemployeer replied:
Same as above

Zummek commented on web-app/src/components/tournament/GridListMobile.vue:
```diff
@@ -0,0 +1,76 @@
+      <template v-slot:item="props">
+        <div class="row q-pa-xs col-12">
+          <q-btn no-caps flat padding="none" class="col-12">
+            <q-card class="row col-12">
+              <q-card-section
+                class="col-6 overflow-hidden"
+                style="text-align:left"
+              >
+                <div class="text-grey-14">
+                  Name:
+                </div>
+                <strong>{{ props.row.name }}</strong>
+              </q-card-section>
+              <q-card-section
+                class="col-6"
+                style="text-align:right"
+                :style="
+                  props.row.status === 'In progress'
+                    ? 'color: orange'
+                    : 'color: green'
+                "
+              >
```
Maybe better make this on divs, not q-card-sections. bcz you have to use overflow-hidden when you use col-6 (1:1) - a lot of space for status name
But this is a details again 
`[[Name, <name>] <status>]`

Zummek commented on web-app/src/pages/tournament/TournamentsList.vue:
```diff
@@ -0,0 +1,89 @@
+<template>
+  <div>
+    <grid-list-mobile
+      :data="data"
+      :columns="columns"
+      :filter="filter"
+      :expanded="expanded"
+      :pagination="pagination"
+    />
+    <grid-list-desktop
+      :data="data"
+      :columns="columns"
+      :filter="filter"
+      :expanded="expanded"
+      :pagination="pagination"
+    />
```
1. Move here screen size condition `v-if ... v-else`, it would be more readable
2. `GridList...` its strange name. `Grid` is not important. I suggest renaming this to `TournamentList...`

eastemployeer replied:
1. It's 100% readable because it is CALLED `grid-list-mobile` and `grid-list-desktop`. 
2. It is coming from tournament folder so it is obvious it is for tournament. GridList makes it more clear what type of list It actually is. No need to overwrite word "Tournament"

Zummek replied:
1. When i open this i didn't know how it's work, when You want to 'run' mobile and desktop in the same time, so here's my suggestion. if condition will tell us more
2. I think grid or not, type of table is not important, anyway ok

Zummek commented on web-app/src/pages/tournament/TournamentsList.vue:
```diff
@@ -0,0 +1,89 @@
+<script lang="ts">
+import { Vue, Component } from 'vue-property-decorator';
+import GridListMobile from '../../components/tournament/GridListMobile.vue';
+import GridListDesktop from '../../components/tournament/GridListDesktop.vue';
+@Component({
+  components: { GridListMobile, GridListDesktop },
+})
+export default class TournamentsList extends Vue {
```
```suggestion
import GridListDesktop from '../../components/tournament/GridListDesktop.vue';

@Component({
  components: { GridListMobile, GridListDesktop },
})
export default class TournamentList extends Vue {
```
Should be `TournamentList`. This same in file name

Zummek reviewed (changes requested):
1. Try to pack this view to box (q-card). On large/super large screens this view is not friendly. For this look at (#14) 
3. Did you try to code one responsive/modifiable list for each device?

Buglik committed:
Added routing
for now to index.vue (there's no creator or details pages in this branch)
meta.public is set for true for now as well

Buglik committed:
Added routing
for now to index.vue (there's no creator or details pages in this branch)
meta.public is set for true for now as well

Buglik committed:
added routing to MainLayout toolbar
changed toolbar to q-tabs (from q-card)

Buglik committed:
added routing to MainLayout toolbar
changed toolbar to q-tabs (from q-card)

Buglik committed:
changed path structure
was /index, /tournamentsList ect.
is /app/index, /app/tournamentsList ect

Buglik committed:
changed path structure
was /index, /tournamentsList ect.
is /app/index, /app/tournamentsList ect

Zummek commented:
Is ready to marge or not? @eastemployeer @Buglik

eastemployeer committed:
Fix routes

eastemployeer committed:
Fix routes

eastemployeer committed:
Fixed routes

eastemployeer committed:
Fixed routes

eastemployeer committed:
Fix enrichTournamentWithMSUsers function types

eastemployeer committed:
Add field isFinished for Tournament

eastemployeer committed:
Update model

eastemployeer committed:
Start connecting api to tournament list view

eastemployeer committed:
Connect to API, fixed pagination

Zummek commented on server/database/models/TournamentModel.ts:
```diff
@@ -28,6 +28,7 @@ const TournamentSchema = new Schema({
   ownerId: Schema.Types.String,
   teams: [{ type: Schema.Types.ObjectId, ref: 'Team' }],
   matches: [MatchSchema],
+  isFinished: { type: Boolean },
```
```suggestion
  isFinished: { type: Schema.Types.Boolean },
```

Zummek commented on web-app/src/components/models.ts:
```diff
@@ -8,6 +8,14 @@ export interface Meta {
+}
+export interface IProps {
+  pagination: IPagination;
 }
```
idk, this is random file `models.ts`, okej, we have types here but.. `IProps` is generic name, but its use is very limited

Zummek commented on web-app/src/pages/tournament/TournamentsList.vue:
```diff
@@ -1,34 +1,39 @@
+    page: 1,
+    rowsPerPage: 1,
+    rowsNumber: 0,
   };
```
I think it may not have a type `IPagination`. Object got a type after assignment

eastemployeer replied:
Bad thinking

Zummek replied:
What do you mean?  TS/Vue should infer a type from assignment

Zummek commented on web-app/src/pages/tournament/TournamentsList.vue:
```diff
@@ -46,50 +51,46 @@ export default class TournamentsList extends Vue {
+  private async pageChanged() {
+    await this.loadData(this.pagination.page);
+  }
+  private async loadData(page: number) {
+    //Funkcja wyciąga wszystkich uczestników danego turnieju.
+    const mapUsers = (teams: Team[]) => {
+      //1 etap - wynik: np. [[{firstName: "Artur", lastName: "Oswald"},{firstName: "Chris", lastName: "Green"}],[...]]
+      const mappedUsers = teams.map((team: Team) => {
+        return team.members.map((user: User) => ({
+          firstName: user.firstName,
+          lastName: user.lastName,
+        }));
+      });
+      //2 etap -> tworzenie jednej tablicy
+      const cleanUsers: ListUser[] = [];
+      mappedUsers.forEach(team => {
+        cleanUsers.push(...team);
+      });
+      return cleanUsers;
+    };
+
+    const tournamentsResponse = await api.tournament.getAllTournaments(page, 1);
+    this.pagination.rowsNumber = tournamentsResponse.totalRows;
+    this.trueData = tournamentsResponse.tournaments.map((el, index) => {
+      const cleanUsers = mapUsers(tournamentsResponse.tournaments[index].teams);
+      return {
+        id: el.id as string,
+        name: el.name,
+        status: el.isFinished === true ? 'Finished' : 'In progress',
+        participants: cleanUsers,
+      };
+    });
+    console.log(this.trueData);
+  }
```
```suggestion
  this.tournaments = await api.tournament.getAllTournaments(page, 1)
```
lol, I think you know what's going on

Zummek replied:
Don't change status/isFinished here for string. Handle it in a template and translate with i18n

Zummek commented on web-app/src/pages/tournament/types.d.ts:
```diff
@@ -0,0 +1,12 @@
+import User from 'app/../shared/types/User';
+
+export interface TournamentListData {
+  id: string;
+  name: string;
+  status: string;
+  participants: ListUser[];
+}
+export interface ListUser {
+  firstName: string;
+  lastName: string;
+}
```
Delete this file :)

eastemployeer replied:
No :)

Zummek commented on web-app/src/router/routes.ts:
```diff
@@ -12,20 +12,20 @@ const routes: RouteConfig[] = [
         name: 'TournamentDetails',
         // component: () => import('pages/tournament/Details.vue'),
         component: () => import('pages/Index.vue'),
         meta: { public: true }, // TODO: change to false
```
Here should be `tournament/:id`

eastemployeer replied:
No it shouldn't

Zummek replied:
maybe `should` is bad word. It was only suggestion. look for example:
`https://github.com/Zummek/best-tournament-app/pull/13`
`https://github.com/Zummek/best-tournament-app/pulls`

Zummek commented on web-app/src/services/API/apiResources/types.ts:
```diff
@@ -29,6 +29,12 @@ export interface getTournamentResponse {
+export interface getTournamentsResponse {
+  data: {
+    totalRows: number;
+    tournaments: Tournament[];
```
This is a beautiful object ready to handle in views

Zummek commented on server/database/repositories/TournamentRepository.ts:
```diff
@@ -41,6 +41,7 @@ function toTournamentDb(t: TournamentWithoutMS) : TournamentDb {
     ownerId: t.ownerId,
     teams: t.teams.map((team) => toTeamDb(team)),
     matches: t.matches.map((match) => toMatchDb(match)),
+    isFinished: t.isFinished,
```
You add a useless object to Tournament. `isFinished` is always false. I think this should be another task. 
Anyway now you have to add logic to updateMatchScore to update tournament.isFinished

eastemployeer replied:
No, it is not useless. We discussed it many times before. It stays here as it is. It's not always false. You can check in google what "is finished" means.

Zummek replied:
I mean this object doesn't have logic. In updateMatchScore You should code condition to set isFinished

Buglik committed:
cleanup
removed trash from mobile view
set search input width in destopGrid as default

Buglik committed:
added locales
and a little cleanup
